File: tutorial5-Restful-services-integration/tutorial5-Restful-services-integration/java/src/main/PicassoBot.java
package main;

import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.InputSource;

import javax.xml.xpath.XPath;
import javax.xml.xpath.XPathConstants;
import javax.xml.xpath.XPathExpressionException;
import javax.xml.xpath.XPathFactory;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URI;
import java.net.URL;

public class PicassoBot {

  // Defines the server address
  static String serverAddress = "http://localhost:8080";

  enum Direction {
    UP("up"), DOWN("down"), RIGHT("right"), LEFT("left");
  
    private String dir;
  
    Direction() {
    }
  
    Direction(String dir) {
      this.dir = dir;
    }
  
    public String getDir() {
      return dir;
    }
  
    public void setDir(String dir) {
      this.dir = dir;
    }
  }
  
  public static void main(String[] args) {
    /*
     * The main function just loops asking input from the user.
     * 
     * tiles
     * => Lists the positions and colors of all tiles that are not empty (have been
     * created)
     * 
     * dot x y r g b
     * => Paints a dot in position (x, y) with the color (r, g, b)
     * 
     * line x y direction length r g b
     * => draws a line of 'length' starting in point (x, y), with direction as 'dir'
     * (RIGHT, LEFT, UP, DOWN) and color (r, g, b)
     * 
     */

    PicassoBot picasso = new PicassoBot();

    BufferedReader in = new BufferedReader(new InputStreamReader(
        System.in));

    String cmd;
    System.out.println("Choose your command:");
    try {
      while ((cmd = in.readLine()) != null) {
        if (cmd.equals("tiles")) {
          picasso.printTiles();

        } else if (cmd.startsWith("dot")) {
          String[] cmdArgs = cmd.split(" ");
          if (cmdArgs.length == 6) {
            int x = Integer.parseInt(cmdArgs[1]);
            int y = Integer.parseInt(cmdArgs[2]);
            int r = Integer.parseInt(cmdArgs[3]);
            int g = Integer.parseInt(cmdArgs[4]);
            int b = Integer.parseInt(cmdArgs[5]);

            boolean res = picasso.paintDot(x, y, r, g, b);
            if (res == true) {
              System.out.println("Painted your dot!");
            }
          }

        } else if (cmd.startsWith("line")) {
          String[] cmdArgs = cmd.split(" ");
          if (cmdArgs.length == 8) {
            int x = Integer.parseInt(cmdArgs[1]);
            int y = Integer.parseInt(cmdArgs[2]);
            Direction dir = Direction.valueOf(cmdArgs[3]);
            int length = Integer.parseInt(cmdArgs[4]);
            int r = Integer.parseInt(cmdArgs[5]);
            int g = Integer.parseInt(cmdArgs[6]);
            int b = Integer.parseInt(cmdArgs[7]);

            boolean res = picasso.paintLine(x, y, dir, length, r, g, b);
            if (res == true) {
              System.out.println("Painted your line!");
            }
          }
        } else {
          System.out.println("Please choose one of the available commands");
        }
        System.out.println("Choose your command:");
      }
    } catch (IOException e) {
      e.printStackTrace();
    }

  }

  private void printTiles() {
    try {
      // Initiate the REST client.
      URL url = URI.create(serverAddress + "/game/board").toURL();
      HttpURLConnection connection = (HttpURLConnection) url.openConnection();

      // HTTP Verb. Get requests data from the server.
      connection.setRequestMethod("GET");

      // We are interested in the output
      connection.setDoOutput(true);

      // If there is a 3xx error, we want to know.
      connection.setInstanceFollowRedirects(false);

      // The Accept header defines what kind of formats we are interested in.
      // You should play with "*/*", "application/xml" and "application/json"
      // JSON might need a third party library to parse the response.
      connection.setRequestProperty("Accept", "application/xml");

      // User Agent is the name of your application.
      // Some of the most common are Mozilla, Internet Explorer and GoogleBot.
      connection.setRequestProperty("User-agent", "Pablo v1");
      
      // If we get a Redirect or an Error (3xx, 4xx and 5xx)
      if (connection.getResponseCode() >= 300) {
        // We want more information about what went wrong.
        debug(connection);
      }

      // Response body from InputStream.
      InputSource inputSource = new InputSource(connection.getInputStream());

      // XPath is a way of reading XML files.
      XPathFactory factory = XPathFactory.newInstance();
      XPath xPath = factory.newXPath();

      // here we are querying the document (much like SQL) for all the item tags
      // inside row elements.
      NodeList nodes = (NodeList) xPath.evaluate("/board/row/item", inputSource, XPathConstants.NODESET);
      // The last argument defines the type of result we are looking for. Might be
      // NODESET for a list of Nodes
      // or NODE for a single node.

      // We don't need the connection anymore once we get the nodes.
      connection.disconnect();

      // Pretty printing of output
      System.out.println("===========================================");
      for (int i = 0; i < nodes.getLength(); i++) {
        Node node = nodes.item(i);

        // Fetching the attributes of the item element
        String created = node.getAttributes().getNamedItem("created").getTextContent();

        // We only want to print tiles that have been created and painted
        if ("true".equals(created)) {
          String x = node.getAttributes().getNamedItem("x").getTextContent();
          String y = node.getAttributes().getNamedItem("y").getTextContent();
          String r = node.getAttributes().getNamedItem("r").getTextContent();
          String g = node.getAttributes().getNamedItem("g").getTextContent();
          String b = node.getAttributes().getNamedItem("b").getTextContent();

          System.out.println(String.format("Tile (%s,%s) with color (%s, %s, %s)", x, y, r, g, b));
        }

      }
      System.out.println("===========================================");

    } catch (IOException e) {
      e.printStackTrace();
    } catch (XPathExpressionException e) {
      e.printStackTrace();
    }
  }

  private void debug(HttpURLConnection connection) throws IOException {
    // This function is used to debug the resulting code from HTTP connections.

    // Response code such as 404 or 500 will give you an idea of what is wrong.
    System.out.println("Response Code:" + connection.getResponseCode());

    // The HTTP headers returned from the server
    System.out.println("_____ HEADERS _____");
    for (String header : connection.getHeaderFields().keySet()) {
      System.out.println(header + ": " + connection.getHeaderField(header));
    }

    // If there is an error, the response body is available through the method
    // getErrorStream, instead of regular getInputStream.
    BufferedReader in = new BufferedReader(new InputStreamReader(
        connection.getErrorStream()));
    StringBuilder builder = new StringBuilder();
    String inputLine;
    while ((inputLine = in.readLine()) != null)
      builder.append(inputLine);
    in.close();
    System.out.println("Body: " + builder);
  }

  private boolean paintDot(int x, int y, int r, int g, int b) {
    try {
      URL url = URI.create(serverAddress + "/game/board/" + x + "/" + y).toURL();
      HttpURLConnection connection = (HttpURLConnection) url.openConnection();

      // Let's try to paint a tile and hope that there is no tile there yet
      // Method is now POST for creating a new tile
      connection.setRequestMethod("POST");

      connection.setDoOutput(true);
      connection.setInstanceFollowRedirects(false);
      connection.setRequestProperty("Accept", "application/xml");
      connection.setRequestProperty("User-agent", "Pablo v1");

      // We can use getOutputStream() for passing the values for r, g and b in the
      // request's body
      OutputStream os = connection.getOutputStream();
      os.write(("r=" + r + "&g=" + g + "&b=" + b).getBytes());
      os.flush();

      if (connection.getResponseCode() >= 300) {
        debug(connection);
      } else {
        return true;
      }

    } catch (IOException e) {
      e.printStackTrace();
    }

    return false;

  }

  private boolean paintLine(int x, int y, Direction direction, int length, int r, int g, int b) {
    int curX = x;
    int curY = y;

    for (int i = 0; i < length; i++) {
      // Paint dot by dot
      boolean res = paintDot(curX, curY, r, g, b);

      if (res == false) {
        // Failed to paint one of the dots. Stop here.
        return false;
      }

      // Move according to the direction
      switch (direction) {
        case UP:
          --curY;
          break;
        case DOWN:
          ++curY;
          break;
        case RIGHT:
          ++curX;
          break;
        case LEFT:
          --curX;
          break;
      }
    }
    return true;
  }
}

File: tutorial5-Restful-services-integration/tutorial5-Restful-services-integration/python/src/PicassoBot.py
import requests
import xml.etree.ElementTree as ET

class Direction:
  UP = "up"
  DOWN = "down"
  RIGHT = "right"
  LEFT = "left"

class PicassoBot:

  #  Defines the server address
  server_address = "http://localhost:8080"

  def main(self):
    """
    The main function loops asking input from the user.
    - `tiles` -> Lists the positions and colors of all tiles that are not empty
    - `dot x y r g b` -> Paints a dot in position (x, y) with the color (r, g, b)
    - `line x y direction length r g b` -> Draws a line starting from (x, y) in the given direction with length
    """

    print("Choose your command:")
    while True:
      try:
        cmd = input().strip()
        if cmd == "tiles":
          self.print_tiles()
        elif cmd.startswith("dot"):
          cmd_args = cmd.split()
          if len(cmd_args) == 6:
            x, y, r, g, b = map(int, cmd_args[1:])
            if self.paint_dot(x, y, r, g, b):
              print("Painted your dot!")
        elif cmd.startswith("line"):
          cmd_args = cmd.split()
          if len(cmd_args) == 8:
            x, y = map(int, cmd_args[1:3])
            direction = cmd_args[3]
            length, r, g, b = map(int, cmd_args[4:])
            if self.paint_line(x, y, direction, length, r, g, b):
              print("Painted your line!")
        else:
          print("Please choose one of the available commands")
      except (ValueError, IndexError):
        print("Please choose one of the available commands")
      except KeyboardInterrupt:
        print("\nExiting...")
        break

  def print_tiles(self):
    """
    This function gets the board information and pretty prints it in our
    console
    """

    try:
      # Set the URL for the request.
      url = f"{self.server_address}/game/board"

      # The Accept header defines what kind of formats we are interested in.
      # You should play with "*/*", "application/xml" and "application/json"
      # JSON might need a third party library to parse the response.
      # User Agent is the name of your application.
      # User Agent is the name of your application.
      # Some of the most common are Mozilla, Internet Explorer and GoogleBot.
      headers = {"Accept": "application/xml", "User-Agent": "Pablo v1"}

      # Sends the REST request and stores the response
      response = requests.get(url, headers=headers, allow_redirects=False)

      # If we get a Redirect or an Error (3xx, 4xx and 5xx)
      if response.status_code >= 300:
        print(f"Error: {response.status_code}")
        self.debug(response)
        return

      # Gets the XML from the response text. 
      # ElementTree is a way of reading XML files.
      root = ET.fromstring(response.text)

      # Always close the connection when it is no longer necessary
      response.close()

      # here we are checking the document for all the item info
      # filtering for each created element and pretty printing it
      print("===========================================")
      for item in root.findall(".//item"):
        if item.get("created") == "true":
          x = item.get("x")
          y = item.get("y")
          r = item.get("r")
          g = item.get("g")
          b = item.get("b")
          print(f"Tile ({x},{y}) with color ({r}, {g}, {b})")
      print("===========================================")

    # In case of any other failure we print the information back
    except requests.RequestException as e:
      print(f"Request failed: {e}")

  def paint_dot(self, x, y, r, g, b):
    """
    This function paints a dot on the board
    """

    try:
      # Sets the URL for our request in the specified coordinates x, y
      url = f"{self.server_address}/game/board/{x}/{y}"
      # Set heards for our request
      headers = {"Accept": "application/xml", "User-Agent": "Pablo v1"}
      # Set the color information for our new tyle
      data = {"r": r, "g": g, "b": b}
      # Send the request to the server
      response = requests.post(url, headers=headers, data=data)


      # If we get anything that is a redirect, client side or service side error
      # we print out the information for debugging
      if response.status_code >= 300:
        self.debug(response)
        response.close()
        return False
      return True

    # If the request fails for any other reason we print out the debugging information
    except requests.RequestException as e:
      print(f"Request failed: {e}")
      response.close()
      return False

  def paint_line(self, x, y, direction, length, r, g, b):
    """
    This function paints a line in the board game
    """
    cur_x, cur_y = x, y

    # In our case we will paint our line dot by dot
    # by printing a dot and the update the current position
    # in the desired direction  
    for _ in range(length):
      # Paints dot in the current position
      if not self.paint_dot(cur_x, cur_y, r, g, b):
        # Fails if a dot failed to print.
        return False

      # We update the position according to the
      # disired line direction 
      if direction == Direction.UP:
        cur_y -= 1
      elif direction == Direction.DOWN:
        cur_y += 1
      elif direction == Direction.RIGHT:
        cur_x += 1
      elif direction == Direction.LEFT:
        cur_x -= 1
    return True

  def debug(self, response):
    """
    This function is used to debug the resulting code from HTTP connections.
    """

    # Response codes such as 404 or 500 will give you an idea of what is wrong.
    print(f"Response Code: {response.status_code}")

    # The HTTP headers returned from the server
    print("_____ HEADERS _____")
    for key, value in response.headers.items():
      print(f"{key}: {value}")
    
    # The HTTP body returned from the server
    print(f"Body: {response.text}")

if __name__ == "__main__":
  PicassoBot().main()
